### PR TITLE
Fixing foreign key constraints error

### PR DESCRIPTION
**Pull request by mgonyan**
this PR fixes https://github.com/doctrine/data-fixtures/issues/230

### CODE REVIEW

**Comment by Ocramius** [comment]
@mgonyan fix seems good, but isn't backed by test scenario...

**Comment by peterrehm** [comment]
Awesome. Was investigating in that direction but then ran out of time.

**Comment by mgonyan** [comment]
Working on it

**Comment by mgonyan** [comment]
@Ocramius I've added the two different test scenario that covers:
1. Checks Node dependencies have not been changed,  testCommitOrder
2. Checks purge remove table in the right order, testPurgeTableCorrectOrder

**Comment by peterrehm** [comment]
It seems the fix needs still some adjustments as I am now getting a `CircularReferenceException` on the purge. This now happens on a simple relation.

File: lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php [hunk 1]
```diff
@@ -152,7 +148,9 @@ private function getCommitOrder(EntityManagerInterface $em, array $classes)
         $sorter = new TopologicalSorter();
 
         foreach ($classes as $class) {
-            $sorter->addNode($class->name, $class);
+            if ( ! $sorter->hasNode($class->name)) {
```
Comment (peterrehm) [inline lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php]: This change causes the CircularReferenceException

**Comment by mgonyan** [comment]
@peterhorne could you please give me more details about or an example what's causing a `CircularReferenceException` in order to add it to the test suite

**Comment by peterrehm** [comment]
I have a Customer, a ContactPerson and a User entity. The relationship is Customer -> Contactperson -> User. I get the `CircularReferenceException` between Customer and User.

**Comment by mgonyan** [comment]
@peterrehm thank you for the example, but I am afraid, I could not find a way to reproduce the issue. I added a test with a similar scenario and it is not failing for me. @Ocramius what do you think?

**Comment by bendavies** [comment]
This PR fixes most of my issues, but I also get a `CircularReferenceException` for my entities. I have a fix however. I'll open a PR based on this one.

**Comment by mgonyan** [comment]
@bendavies  could you please tell me how your entities look like and their relationships in order to have a solid test, adjust this PR, thank you

**Comment by bendavies** [comment]
sure you need this:

Single table inheritance:
Parent -> Child

Parent ManyToOne Parent
Parent ManyToOne Organisation

That should get you the circular exception,

then to fix it, at https://github.com/mgonyan/data-fixtures/blob/63c795a2591e2f54bfaefe6ee8de5f8f85849133/lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php#L147
```
if (isset($association['inherited'])) {
    continue;
}
```

**Comment by peterrehm** [comment]
@mgonyan Can only look at it in detail on monday.

**Comment by mgonyan** [comment]
Thank you @peterrehm and @bendavies, i've been able to reproduce the issue adding a simple table inheritance as you pointed out, please have a look at it a tell me what you guys think.

**Comment by bendavies** [comment]
nice one!

**Comment by bendavies** [comment]
PR looks good. has fixed all my issues.

**Comment by peterrehm** [comment]
Looks like it does not yet fix my inheritance issue. I am having a Single-Table-Inheritance of BaseUser -> and User. And then the relation is as mentioned above. Customer -> ContactPerson -> User. I am getting CircularReference errors between Customer and User.

**Comment by mikeSimonson** [comment]
@peterhorne Can you make a PR with a failing test ?

**Comment by mgonyan** [comment]
@mikeSimonson I have a test that makes the PR failed, However, I am not sure how to solve that issue,
EntityA -> EntityB -> EntntyC -> EntityA, I'll push it later

**Comment by peterrehm** [comment]
@mgonyan Any updates?

**Comment by mikeSimonson** [comment]
@mgonyan Can you push your test ?

**Comment by mgonyan** [comment]
@mikeSimonson @peterrehm sorry for the delay, can you please guys have a look at the changes. I am still not sure about the circular reference issue with single table reference.

**Comment by mikeSimonson** [comment]
@mgonyan The 2 tests that fail now illustrate the BC break right ?
[testCommitOrder](https://github.com/doctrine/data-fixtures/pull/231/files#diff-50fdd27b992ea6b371c298ce334e420fR69) and [testPurgeTableCorrectOrder](https://github.com/doctrine/data-fixtures/pull/231/files#diff-50fdd27b992ea6b371c298ce334e420fR83).

**Comment by mgonyan** [comment]
@mikeSimonson yes I do, if this is the bug which @peterrehm mentioned before I will try to fixing it asap However cycles are possible due to a bad Entity Relation design or particular business requirement. I think this PR https://github.com/doctrine/data-fixtures/pull/233 could help to give us a clue where it is the cycle and rethink the relation model.

**Comment by mikeSimonson** [comment]
@mgonyan Thanks for your help

**Comment by mgonyan** [comment]
thank u too. @mikeSimonson , WDYT @peterrehm  ?

**Pull request closed without merging**